### ISSUE TITLE
convolutional autoencoder MNIST

### ISSUE THREAD

**Issue by Harhiem** [rstudio/keras#144]
Dear all,

I am struggling with creating in R the convolution autoencoder for the MNIST dataset as described in https://blog.keras.io/building-autoencoders-in-keras.html.

This is my R code based on the blog:

### convolution autoencoder for MNIST dataset ###
library(keras)

batch_size <- 128
img_rows <- 28
img_cols <- 28

mnist <- dataset_mnist()
x_train <- mnist$train$x
y_train <- mnist$train$y
x_test <- mnist$test$x
y_test <- mnist$test$y

dim(x_train) <- c(nrow(x_train), img_rows, img_cols, 1) 
dim(x_test) <- c(nrow(x_test), img_rows, img_cols, 1)
input_dim <- c(img_rows, img_cols, 1)
input_dim

x_train <- x_train / 255
x_test <- x_test / 255

### define model
CAE <- keras_model_sequential()
CAE %>%
  layer_conv_2d(filters = 16, kernel_size = c(3,3), activation = 'relu',padding='same', input_shape = input_dim)%>% 
  layer_max_pooling_2d(pool_size = c(2, 2),padding='same')%>% 
  layer_conv_2d(filters = 8, kernel_size = c(3,3), activation = 'relu',padding='same')%>% 
  layer_max_pooling_2d(pool_size = c(2, 2),padding='same')%>% 
  layer_conv_2d(filters = 8, kernel_size = c(3,3), activation = 'relu',padding='same')%>% 
  layer_max_pooling_2d(pool_size = c(2, 2),padding='same')%>% 
  layer_conv_2d(filters = 8, kernel_size = c(3,3), activation = 'relu',padding='same')%>% 
  layer_upsampling_2d(size = c(2, 2))%>%
  layer_conv_2d(filters = 8, kernel_size = c(3,3), activation = 'relu',padding='same')%>% 
  layer_upsampling_2d(size = c(2, 2))%>%
  layer_conv_2d(filters = 16, kernel_size = c(3,3), activation = 'relu',padding='same')%>% 
  layer_upsampling_2d(size = c(2, 2))%>%
  layer_conv_2d(filters = 1, kernel_size = c(3,3), activation = 'sigmoid',padding='same') 
  
  
summary(CAE)

### compile model
CAE %>% compile(optimizer='adadelta', loss='binary_crossentropy')

CAE %>% fit(
    x_train, x_train,
    epochs = 50,
    batch_size = 128,
    shuffle = TRUE,
    validation_data = list(x_test, x_test)
  )
  
The main problem is to get the dimensions correct as is shown by summary(CAE) 
> summary(CAE) 
> summary(CAE)
__________________________________________________________________________________________________________________________
Layer (type)                                          Output Shape                                     Param #            
==========================================================================================================================
conv2d_1 (Conv2D)                                     (None, 28, 28, 16)                               160                
__________________________________________________________________________________________________________________________
max_pooling2d_1 (MaxPooling2D)                        (None, 14, 14, 16)                               0                  
__________________________________________________________________________________________________________________________
conv2d_2 (Conv2D)                                     (None, 14, 14, 8)                                1160               
__________________________________________________________________________________________________________________________
max_pooling2d_2 (MaxPooling2D)                        (None, 7, 7, 8)                                  0                  
__________________________________________________________________________________________________________________________
conv2d_3 (Conv2D)                                     (None, 7, 7, 8)                                  584                
__________________________________________________________________________________________________________________________
max_pooling2d_3 (MaxPooling2D)                        (None, 4, 4, 8)                                  0                  
__________________________________________________________________________________________________________________________
conv2d_4 (Conv2D)                                     (None, 4, 4, 8)                                  584                
__________________________________________________________________________________________________________________________
up_sampling2d_1 (UpSampling2D)                        (None, 8, 8, 8)                                  0                  
__________________________________________________________________________________________________________________________
conv2d_5 (Conv2D)                                     (None, 8, 8, 8)                                  584                
__________________________________________________________________________________________________________________________
up_sampling2d_2 (UpSampling2D)                        (None, 16, 16, 8)                                0                  
__________________________________________________________________________________________________________________________
conv2d_6 (Conv2D)                                     (None, 16, 16, 16)                               1168               
__________________________________________________________________________________________________________________________
up_sampling2d_3 (UpSampling2D)                        (None, 32, 32, 16)                               0                  
__________________________________________________________________________________________________________________________
conv2d_7 (Conv2D)                                     (None, 32, 32, 1)                                145                
==========================================================================================================================
Total params: 4,385
Trainable params: 4,385
Non-trainable params: 0
__________________________________________________________________________________________________________________________

Any suggestions to improve the code is appreciated.

With kind regards,

Harry

**Comment by jjallaire** [comment]
You need to leave the padding argument off of the second to last convolution (that's what the referenced Python example does as well). Here is the required change: 

https://gist.github.com/jjallaire/1fa0256704692dd2d470986f2547d8f5/revisions
